File: py_implementation/BayesClassifier.py
import numpy as np

# class implementing a Naive Bayes Classifier
class BayesClassifier:
	
	def __init__(self, wanted):
		self.wanted			= wanted
		self.num_dimensions = []
		self.num_classes	= []
		self.histos 		= []
		self.priors			= []
		self.num_bins   	= []
		self.val_ranges		= []
		self.bin_widths		= []
		self.irrelevant_dims = []
		self.trained		= False
		self.CDF			= []

	# get the optimal number of bins for d-th dimension
	# using the Freedman Diaconis rule
	def get_optimal_bins(self, data, d):
		N = data.shape[0]
		# calculate the IQR
		q75, q25 = np.percentile(data[:,d], [75,25])
		iqr = q75 - q25
		# optimal width according to Freedman Diaconis
		width = 2 * iqr / N**(1./3)
		# if the width is 0, this dimension is irrelevant!
		if width == 0.0:
			self.irrelevant_dims.append(d)
			self.val_ranges.append(0.)
			self.num_bins.append(0.)
			self.bin_widths.append(0.)
			return 0
		# calculate the number of bins
		value_range = np.max(data[:,d]) - np.min(data[:,d])
		bins = int(round(value_range / width))
		
		# limit the number of bins
		if bins > N/4:
			bins = N/4
			width = value_range / bins
		
		self.val_ranges.append(value_range)
		self.num_bins.append(bins)
		self.bin_widths.append(width)
		
	
	# train the classifier by building the histograms
	def train(self, train_data, train_label): 
		if self.trained:
			print "Retraining Bayes Classifier with new data!"

		#number of classes
		self.num_classes = len(self.wanted)
		#number of dimensions
		self.num_dimensions = train_data.shape[1]
		
		print "Dimensions:", self.num_dimensions
		# determine the number of bins and the bin width for each dimensions individually
		self.num_bins 	= []
		self.val_ranges = []
		self.bin_widths = []
		for d in range(self.num_dimensions):
			bins = self.get_optimal_bins(train_data,d)
	
		#initialise histogram for each dimension
		self.histos = []
		for d in range(self.num_dimensions):
			self.histos.append( np.zeros( (self.num_bins[d], self.num_classes) ) )
		
		# count entries in histograms
		for i in range(train_data.shape[0]):
			for d in range(self.num_dimensions):
				if d in self.irrelevant_dims:
					continue
				else:
					val = train_data[i][d]
					# assumung that val_min = 0
					curr_bin = int( (val / self.val_ranges[d]) * self.num_bins[d] )
					# curr_bins == num_bins can happen for the maximal value
					if curr_bin == self.num_bins[d]:
						curr_bin = self.num_bins[d] - 1
					label = train_label[i]
					#get the correct class
					cla = np.searchsorted(self.wanted,label)
					assert(cla < self.num_classes)
					self.histos[d][curr_bin,cla] += 1
			
		# normalize the histograms
		N_class = []
		for c in range(self.num_classes):
			clas = self.wanted[c]
			num = np.where(train_label == clas)[0].shape[0]
			N_class.append(num)

		for d in range(self.num_dimensions):
			for c in range(self.num_classes):
				if d in self.irrelevant_dims:
					continue
				else:
					self.histos[d][:,c] /= N_class[c]# * self.bin_widths[d])
		
		#compute the priors
		self.priors = []
		n_samples = train_label.shape[0]
		for c in range(self.num_classes):
			self.priors.append(N_class[c] / float(n_samples))
	
		self.trained = True
	

	# classify test_data
	def classify(self, test_data):
		
		if not self.trained:
			raise RuntimeError("BayesClassifier: classify called, before calling train!")

		results = []
		for i in range(test_data.shape[0]):
			probs = []
			#iterate over the classes to get the individual prob
			for c in range(self.num_classes):
				# calculate the likelihoods
				likelihood = 1.
				for d in range(self.num_dimensions):
					if d in self.irrelevant_dims:
						continue
					else:
						val = test_data[i][d]
						curr_bin = int( (val / self.val_ranges[d]) * self.num_bins[d] )
						# this may happen if test data values are bigger than train data vals
						if curr_bin >= self.num_bins[d]:
							curr_bin = self.num_bins[d] - 1
						likelihood *= self.histos[d][curr_bin][c]
				probs.append(likelihood*self.priors[c])
			
			clas = np.argmax(probs)
			results.append(self.wanted[clas])
		
		return np.array(results)

	
	#generate N instances of class C
	def generate(self, N, C):
		if not self.trained:
			raise RuntimeError("BayesClassifier: generate called, before calling train!")
		
		np.random.seed()
		
		C = np.searchsorted(self.wanted,C)

		generated = []
		for _ in range(N):
			gen = np.zeros(self.num_dimensions)
			# for every feature dimension sample from the corresponding histogram
			for d in range(self.num_dimensions):
				
				if d in self.irrelevant_dims:
					gen[d] = 0.
					continue
				else:
					#sampling strategy: sample from the 2 most likely histogram bins, irresponsible of their prob.
					likelihoods = self.histos[d][:,C]
					sorted_bins = np.argsort(likelihoods)
					which_bin   = np.random.randint(3)
					bin_chosen  = sorted_bins[which_bin]
					#sample the pixel value uniformly inside the chosen bin:
					bin_min = bin_chosen*self.bin_widths[d]
					bin_max = (bin_chosen + 1)*self.bin_widths[d]
					dim_val = np.random.uniform(bin_min,bin_max)
					gen[d] = dim_val
			
			generated.append(gen)
		return np.array(generated)

	
	# comute the CDF of the histograms
	def compute_CDF(self): 
		if not self.trained:
			raise RuntimeError("BayesClassifier: generate called, before calling train!")
		
		self.CDF = []
		print len(self.histos)
		print self.histos[0].shape
		
		for d in range(self.num_dimensions):
			if d in self.irrelevant_dims:
				self.CDF.append(0)	
				continue
			
			else:
				histo 	= self.histos[d]
				bins_d	= self.num_bins[d]
				assert(bins_d == histo.shape[0])
				CDF_d 	= np.zeros( (bins_d,self.num_classes) )
				for C in range(self.num_classes):
					CDF_d[0,C] = histo[0,C]
					for b in range(1,bins_d):
						CDF_d[b,C] = CDF_d[b-1,C] + histo[b,C]
				
				self.CDF.append(CDF_d)
	
	
	# compute the inverse CDF-value of x
	def inverse_CDF(self, x, clas):
		if not self.trained:
			raise RuntimeError("BayesClassifier: generate called, before calling train!")
		
		C = np.searchsorted(self.wanted, clas)
		
		assert(x.shape[0] == self.num_dimensions)

		returnval = np.zeros( (self.num_dimensions) )
		
		for d in range(self.num_dimensions):
			if d in self.irrelevant_dims:
				returnval[d] = 0.
				continue
			else:
				CDF_d = self.CDF[d][:,C]
				val = x[d]
				
				# find the bin we are in
				for j in range(CDF_d.shape[0]):
					if val < CDF_d[j]:
						break
				
				# return the position (middle of the bin)
				pos = self.bin_widths[d] * (j + 0.5)
				returnval[d] = pos
		
		return returnval

File: c++_implementation/source/BayesClassifier.cpp
#include <algorithm>
#include <chrono>
#include <exception>
#include <iostream>
#include <iterator>
#include <math.h>
#include <random>

#include "BayesClassifier.h"

using namespace boost::numeric::ublas;

BayesClassifier::BayesClassifier() : 	trained(false),
   										num_instances(0),
										num_dimensions(0),
										num_classes(0),
										instances_per_class(),
										priors(),
										bins(),
										irrelevant_dims(),
										histograms(),
										cdfs()
{}

void BayesClassifier::train(const image_data_t & train_data, const label_data_t & train_label)
{
	if( trained )
	{
		std::cout << "Called BayesClassifier::train with trained = true, retraining the classifier." << std::endl;
	}
// determine number of instances, dimensions and classes
	num_instances	= train_data.size1();
	num_dimensions 	= train_data.size2(); 
// get number of classes, here we assume that class labeling starts at zero and there are no intermediate numbers missing !!!
	auto min_max	= std::minmax_element( train_label.begin(), train_label.end() );
	num_classes    	= (*min_max.second - *min_max.first) + 1;
// calculate the number of class instances and the priors
	instances_per_class.clear();
	priors.clear();
	for( size_t c = 0; c < num_classes; c++)
	{
		size_t N_class = std::count( train_label.begin(), train_label.end(), c);
	  	instances_per_class.push_back(N_class);
		priors.push_back( N_class/static_cast<double>(num_instances) );
	}
// get the optimal bin width for each dimensions
	bins.clear();
	irrelevant_dims.clear();
	for( size_t d = 0; d < num_dimensions; d++ )
	{
		bin_t current_bin = get_optimal_bins(train_data,d);
// debug output
                //std::cout << "dimension: " << d << "\twidth: " << current_bin.width << "\tnumber: " << current_bin.num_bins << std::endl;
		if( current_bin.width == 0.0)
		{
			irrelevant_dims.push_back(d); 
		}
		bins.push_back(current_bin);
	}
// initialize the histograms
	histograms.clear(); 
	for( size_t d = 0; d < num_dimensions; d++)
	{
		histogram_t histo_dim = histogram_t(bins[d].num_bins, num_classes);
		histograms.push_back(histo_dim);
	}
// calculate the histograms from the data
	for( size_t d = 0; d < num_dimensions; d++)
	{
// check if this dimension is irrelevant and continue if it is
		if( std::find( irrelevant_dims.begin(), irrelevant_dims.end(), d) != irrelevant_dims.end() )
		{
			continue;
		}
		else
		{
// iterate over the training data and assign to corresponding bin
			for( size_t i = 0; i < num_instances; i++)
			{
				double val 	= train_data(i,d);
				short label = train_label[i];
				// find the bin
				size_t bin = static_cast<size_t>( ( (val - bins[d].lowest_val) / bins[d].val_range) * bins[d].num_bins );
// lower bin by 1 if it is too big (this may happen...) FIXME
				if( bin == bins[d].num_bins)
				{
					bin--;
				}
// debug output
				//std::cout << "Dim " << d << " instance " << i << " bin " << bin << " bin-max " << bins[d].num_bins << std::endl;
				histograms[d](bin,label) += 1;
			}	
// normalise the histogram
			for( size_t c = 0; c < num_classes; c++)
			{
				row( histograms[d], c ) /= instances_per_class[c]; // TODO also divide by bin width ???
			}
		}
	}
// create the cumulative densities
	compute_cdf();
// set train flag to true
	trained = true;
}

label_data_t BayesClassifier::predict(const image_data_t & test_data)
{
	if( !trained )
	{
		throw std::runtime_error("BayesClassifier::predict trying to predict without having trained!");
	}
	label_data_t labels_return;
// iterate over test instances
	for( size_t i = 0; i < test_data.size1(); ++i)
	{
// calculate the likelihood for each class
		std::vector<double> class_likelihoods;
		for( size_t c = 0; c < num_classes; ++c)
		{
			double likelihood = 1.0;
			for( size_t d = 0; d < num_dimensions; d++)
			{
				double val = test_data(i,d);
// check if this dimension is irrelevant and continue if it is
				if( std::find( irrelevant_dims.begin(), irrelevant_dims.end(), d) != irrelevant_dims.end() )
				{
					continue;
				}
				else
				{
					size_t bin = static_cast<size_t>( ( (val - bins[d].lowest_val) / bins[d].val_range) * bins[d].num_bins );
// set bin to binmax if it is bigger than binmax, this may happen because we havent seen the test_data yet
					if( bin >= bins[d].num_bins)
					{
						bin = bins[d].num_bins - 1;
					}
					likelihood *=  histograms[d](bin,c);
				}
			}
// multiply with the prior
			likelihood *= priors[c];
			class_likelihoods.push_back(likelihood);
		}
// find the class with biggest likelihood
		auto max = std::max_element(class_likelihoods.begin(), class_likelihoods.end());
		size_t max_class = std::distance(class_likelihoods.begin(), max);
		labels_return.push_back(max_class);
	}
	return labels_return;
}

image_data_t BayesClassifier::generate(const size_t N, const short label)
{
	if( !trained )
	{
		throw std::runtime_error("BayesClassifier::generate: Trying to generate without having trained!");
	}
	image_data_t data_return(N,num_dimensions);
// instantiate and seed random generator
	unsigned int seed = std::chrono::system_clock::now().time_since_epoch().count();
	std::default_random_engine gen(seed);
// instantiate uniform real distribution (0 to 1)
	std::uniform_real_distribution<double> distr(0.0,1.0);
// generate N new instances
	for( size_t i = 0; i < N; ++i)
	{
		for( size_t d = 0; d < num_dimensions; d++ )
		{
// check if this dimension is irrelevant and set data to zero if it is
			if( std::find( irrelevant_dims.begin(), irrelevant_dims.end(), d) != irrelevant_dims.end() )
			{
				data_return(i,d) = 0.;
			}
// sampling strategy: sample from the 2 most likely histogram bins with their prob.
// then sample uniform inside the chosen bin
			else
			{
// get data and copy it to safely operate on it
				matrix_column<matrix<double> const> histo(histograms[d], label);
				assert(histo.size() == bins[d].num_bins);
				vector<double> histo_sort( histo.size() );
				std::copy( histo.begin(), histo.end(), histo_sort.begin() );
// sort bins by their probability
				std::sort( histo_sort.begin(), histo_sort.end() );
// get probabilities of the 2 bins with highest probability
				double p = *(histo_sort.end() - 1);
				double q = *(histo_sort.end() - 2);
// normalise the probability
				double p_1 = p / (p + q);
// decide which bin according to the probs
				size_t bin_chosen = 0;
				if( p_1 > distr(gen) )
				{
					auto search_bin = std::find(histo.begin(), histo.end(), p);
					bin_chosen 		= std::distance(histo.begin(), search_bin);
				}
				else
				{
					auto search_bin = std::find(histo.begin(), histo.end(), q);
					bin_chosen 		= std::distance(histo.begin(), search_bin);
				}
// sample uniform inside the bin
				double bin_min = bin_chosen * bins[d].width + bins[d].lowest_val;	
				double bin_max = (bin_chosen + 1) * bins[d].width + bins[d].lowest_val;	
				std::uniform_real_distribution<double> distr_bin(bin_min, bin_max);
				data_return(i,d) = distr_bin(gen);
			}
		}	
	}
	return data_return;
}

vector<double> BayesClassifier::inverse_cdf(const matrix_row<matrix<double> > & data, const short label)
{
	if( !trained )
	{
		throw std::runtime_error("BayesClassifier::inverse_cdf: Trying to calculate the inverse cdf without having trained!");
	}	
	assert(data.size() == num_dimensions);
	vector<double> data_return(num_dimensions);
// iterate over the dimensions
	for( size_t d = 0; d < num_dimensions; d++)
	{
// check if this dimension is irrelevant and set data to zero if it is
		if( std::find( irrelevant_dims.begin(), irrelevant_dims.end(), d) != irrelevant_dims.end() )
		{
			data_return(d) = 0.;
		}
		else
		{
// get the CDF of this dimension and class
			const std::vector<double> & cdf_dim = cdfs[d][label]; 
			double val_dim = data(d);
// find the bin that val_dim is in	
			size_t b = 0;
			while( b < cdf_dim.size() )
			{
				if( val_dim < cdf_dim[b] )
				{
					break;
				}
				b++;
			}
// the value of the inverse cdf corresponds to the position of the middle of the bin
// TODO + 0.5 or - 0.5 ??? I think it is correct this way, but not quite sure
			data_return(d) = bins[d].width * (b - 0.5);
		}
	}
	return data_return;
}
	
double BayesClassifier::get_likelihood(const matrix_row<matrix<double> const> & data, const short label)
{
	if( !trained )
	{
		throw std::runtime_error("BayesClassifier::get_likelihood: called before calling train!");
	}	
	double likelihood = 1.0;
	for( size_t d = 0; d < num_dimensions; d++)
	{
		double val = data(d);
// check if this dimension is irrelevant and continue if it is
		if( std::find( irrelevant_dims.begin(), irrelevant_dims.end(), d) != irrelevant_dims.end() )
		{
			continue;
		}
		else
		{
			size_t bin = static_cast<size_t>( ( (val - bins[d].lowest_val) / bins[d].val_range) * bins[d].num_bins );
// set bin to binmax if it is bigger than binmax, this may happen because we havent seen the test_data yet
			if( bin >= bins[d].num_bins)
			{
				bin = bins[d].num_bins - 1;
			}
			likelihood *=  histograms[d](bin,label);
		}
	}
	return likelihood;
}
	
vector<double> BayesClassifier::get_cdf(const matrix_row<matrix<double> const> & data, const short label)
{
	if( !trained )
	{
		throw std::runtime_error("BayesClassifier::get_cdf: called before calling train!");
	}	
	assert(data.size() == num_dimensions);
	vector<double> cdf_return(num_dimensions);
	for( size_t d = 0; d < num_dimensions; d++)
	{
		if( std::find( irrelevant_dims.begin(), irrelevant_dims.end(), d) != irrelevant_dims.end() )
		{
			cdf_return(d) = 0.;
		}
		else
		{
			double val = data(d);
			size_t bin = static_cast<size_t>( ( (val - bins[d].lowest_val) / bins[d].val_range) * bins[d].num_bins );
// set bin to binmax if it is bigger than binmax, this may happen because we havent seen the test_data yet
			if( bin >= bins[d].num_bins)
			{
				bin = bins[d].num_bins - 1;
			}
			cdf_return(d) = cdfs[d][label][bin];	
		}
	}
	return cdf_return;
}

void BayesClassifier::compute_cdf()
{
	cdfs.clear();
	for(size_t d = 0; d < num_dimensions; d++)
	{
		cdf_t cdf_d;	
		if( std::find( irrelevant_dims.begin(), irrelevant_dims.end(), d) != irrelevant_dims.end() )
		{
			for(size_t c = 0; c < num_classes; c++)
			{
				cdf_d.push_back( std::vector<double>{{0}} );
			}
		}
		else
		{
			const auto & histo 	= histograms[d];
		   	size_t n_bins 		= bins[d].num_bins;
			assert(histo.size1() == n_bins);
			for(size_t c = 0; c < num_classes; c++)
			{
				std::vector<double> cdf_c;
				cdf_c.push_back( histo(0,c) );
				for(size_t b = 1; b < n_bins; b++)
				{
					double val = cdf_c[b-1] + histo(b,c);
					cdf_c.push_back(val);	
				}	
				cdf_d.push_back(cdf_c);
			}
		}
		cdfs.push_back(cdf_d);
	}
}

// TODO what to do for too small bins <-> too many bins
bin_t BayesClassifier::get_optimal_bins(const image_data_t & train_data, const size_t dim)
{
	bin_t bin;
// calculate the optimal bin width with IRQ-criterion
// first get the correct dimension == column
	matrix_column<matrix<double> const> data_dim_aux(train_data, dim);
// copy all data to a vector to safely manipulate it
	vector<double> data_dim( data_dim_aux.size() );
	std::copy(data_dim_aux.begin(), data_dim_aux.end(), data_dim.begin() );
// sort the data in this dimension
	std::sort(data_dim.begin(), data_dim.end());
// get lower quartile
	size_t size    	= num_instances;
	size_t ind_1   	= size / 4;
	double q_1     	= (size%2 == 0) ? ( data_dim(ind_1+1) + data_dim(ind_1) )/2 : data_dim(ind_1);
// get upper quartile
	size_t ind_3   	= 3*size / 4;
	double q_3     	= (size%2 == 0) ? ( data_dim(ind_3+1) + data_dim(ind_3) )/2 : data_dim(ind_3+1);
// calculate the iqr and bin width
	double iqr	   	= q_3 - q_1;
	bin.width 		= 2.*iqr / std::cbrt(num_instances);
// get lowest and highest value in the dimension
	bin.lowest_val  = *( data_dim.begin() );
	bin.highest_val = *( data_dim.end() - 1 );
// get number of bins 
	bin.val_range   = bin.highest_val - bin.lowest_val;
	bin.num_bins 	= static_cast<size_t>(bin.val_range / bin.width);
//limit number of bins
	if(bin.num_bins > std::sqrt(num_instances) )
	{
		std::cout << "Number of bins " << bin.num_bins << " bigger than sqrt(N) = " << std::sqrt(num_instances) << std::endl;
		std::cout << "IQR " << iqr << " N^1/3 " << std::cbrt(num_instances) << std::endl; 
		bin.num_bins = static_cast<size_t>( std::sqrt(num_instances) );
		bin.width    = bin.val_range / bin.num_bins;
		std::cout << "Changing number of bins to: " << bin.num_bins << std::endl;
	}
	return bin;
}
